contracts/RandomBeaconHistory.cdc

```diff
@@ -0,0 +1,167 @@
+                RandomBeaconHistory.lowestHeight = currentBlockHeight
+            }
+
+            RandomBeaconHistory.randomSourceHistory.append(randomSourceHistory)
+        }
+    }
+
+    /* --- RandomSourceHistory --- */
+    //
+    /// Represents a random source value for a given block height
+    ///
+    access(all) struct RandomSource {
+        access(all) let blockHeight: UInt64
+        access(all) let value: [UInt8]
+    
+        init(blockHeight: UInt64, value: [UInt8]) {
+            self.blockHeight = blockHeight
+            self.value = value
+        }
+    }
+
+    /* --- RandomSourceHistoryPage --- */
+    //
+    /// Contains RandomSource values ordered chronologically according to associated block height
+    ///
+    access(all) struct RandomSourceHistoryPage {
+        access(all) let page: UInt64
+        access(all) let perPage: UInt64
+        access(all) let totalLength: UInt64
+        access(all) let values: [RandomSource]
+    
+        init(page: UInt64, perPage: UInt64, totalLength: UInt64, values: [RandomSource]) {
+            self.page = page
+            self.perPage = perPage
+            self.totalLength = totalLength
+            self.values = values
+        }
+    }
+
+    /* --- Contract Methods --- */
+    //
+    /// Getter for the source of randomness at a given block height. Panics if the requested block height either
+    /// precedes or exceeds the recorded history. Note that a source of randomness for block n will not be accessible
+    /// until block n+1.
+    ///
+    /// @param atBlockHeight The block height at which to retrieve the source of randomness
+    ///
+    /// @return The source of randomness at the given block height as RandomSource struct
+    ///
+    access(all) fun sourceOfRandomness(atBlockHeight: UInt64): RandomSource {
+        pre {
+            self.lowestHeight != nil: "History has not yet been initialized"
+            atBlockHeight >= self.lowestHeight!: "Requested block height precedes recorded history"
+            atBlockHeight < getCurrentBlock().height: "Source of randomness not yet recorded"
+        }
+        let index: UInt64 = atBlockHeight - self.lowestHeight!
+        assert(
+            index >= 0 && index < UInt64(self.randomSourceHistory.length),
+            message: "Problem finding random source history index"
+        )
+        return RandomSource(blockHeight: atBlockHeight, value: self.randomSourceHistory[index])
+    }
+
+    /// Retrieves a page from the history of random sources, ordered chronologically
+    ///
+    /// @param page: The page number to retrieve, 0-indexed
+    /// @param perPage: The number of random sources to include per page
+    ///
+    /// @return A RandomSourceHistoryPage containing RandomSource values in choronological order according to
+    /// associated block height
+    ///
+    access(all) view fun getRandomSourceHistoryPage(page: UInt64, perPage: UInt64): RandomSourceHistoryPage {
```

Comment:
Re: the `view` modifier: Is it intentional that the contract can only be used in Cadence 1.0?

Reply:
I missed that keyword. 
The contract was successfully tested by bootstrapping it on current flow go master: https://github.com/onflow/flow-go/pull/4582. How is that possible 🤔 ? Is this a valid contract in current cadence?

Reply:
I was confused for a moment why this was not a parsing error, but after re-familiarizing, it turns out that the current (Pre-Cadence 1.0) parser has a bug where it simply ignores unknown keywords. This is going to be fixed in Cadence 1.0 (through https://github.com/onflow/cadence/pull/2231).

I guess the keywords can be kept for now, given that they are ignored, and as a side-effect the code is already prepped for Cadence 1.0 😅